diff --git a/src/event/ngx_event_openssl.c b/src/event/ngx_event_openssl.c
index 1b789e6..9e500c7 100644
--- a/src/event/ngx_event_openssl.c
+++ b/src/event/ngx_event_openssl.c
@@ -1088,6 +1088,14 @@ ngx_ssl_handshake(ngx_connection_t *c)
     int        n, sslerr;
     ngx_err_t  err;
 
+    struct timeval time;
+    if (c->ssl->handshake.processing == 0) {
+        c->ssl->handshake.processing = 1;
+        ngx_gettimeofday(&time);
+        c->ssl->handshake.start_sec = time.tv_sec;
+        c->ssl->handshake.start_msec = time.tv_usec / 1000;
+    }
+
     ngx_ssl_clear_error(c->log);
 
     n = SSL_do_handshake(c->ssl->connection);
@@ -1153,6 +1161,11 @@ ngx_ssl_handshake(ngx_connection_t *c)
 
         c->ssl->handshaked = 1;
 
+        c->ssl->handshake.processing = 0;
+        ngx_gettimeofday(&time);
+        c->ssl->handshake.end_sec = time.tv_sec;
+        c->ssl->handshake.end_msec = time.tv_usec / 1000;
+
         c->recv = ngx_ssl_recv;
         c->send = ngx_ssl_write;
         c->recv_chain = ngx_ssl_recv_chain;
diff --git a/src/event/ngx_event_openssl.h b/src/event/ngx_event_openssl.h
index 08eff64..e5abbe5 100644
--- a/src/event/ngx_event_openssl.h
+++ b/src/event/ngx_event_openssl.h
@@ -43,6 +43,13 @@ typedef struct {
     size_t                      buffer_size;
 } ngx_ssl_t;
 
+typedef struct {
+    ngx_uint_t                  processing;
+    time_t                      start_sec;
+    ngx_msec_t                  start_msec;
+    time_t                      end_sec;
+    ngx_msec_t                  end_msec;
+} handshake_t;
 
 typedef struct {
     ngx_ssl_conn_t             *connection;
@@ -62,6 +69,8 @@ typedef struct {
     unsigned                    no_wait_shutdown:1;
     unsigned                    no_send_shutdown:1;
     unsigned                    handshake_buffer_set:1;
+
+    handshake_t                 handshake;
 } ngx_ssl_connection_t;
 
 
